### PR TITLE
Ephemeral annotation

### PR DESCRIPTION
**Pull request by ralucaportase**
* add handler for tap 
* move highlight functions from jquery.flot.js to it's own file
* generate plothover event for tap on point
* generate plothovercleanup event for touch when touch event is not tap

### CODE REVIEW

File: jquery.flot.hover.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+            plot.hooks.drawOverlay.push(drawOverlay);
+        }
+
+        var triggerCleanupEvent = function(e) {
+            plot.unhighlight();
+            plot.getPlaceholder().trigger(new CustomEvent('plothovercleanup', { detail: e }));
+        }
+    }
+
+    function onMouseMove(e) {
```
Comment (cristiingineru) [inline jquery.flot.hover.js]: For consistency with the touch event handler(s), these mouse related handlers should be grouped in their own object, or the touch event handler(s) moved outside their {}.

File: jquery.flot.hover.js [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+                // was fixed somewhere around 1.3.x.  We can return to using
+                // .mouseleave when we drop support for 1.2.6.
+
+                eventHolder.bind("mouseleave", onMouseLeave);
+            }
+        }
+
+        function shutdown(plot, eventHolder) {
+            eventHolder[0].removeEventListener('tap', tap.generatePlothoverEvent);
+            eventHolder[0].addEventListener('tap', triggerCleanupEvent);
```
Comment (cristiingineru) [inline jquery.flot.hover.js]: Remove instead of adding the event handler.

File: jquery.flot.hover.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+(function($) {
+    'use strict';
+
+    var options = {
+            grid: {
+                hoverable: true,
+                clickable: true
+            }
+        },
+        plot, highlights = [];
```
Comment (cristiingineru) [inline jquery.flot.hover.js]: These variables are shared among all plots. The right pattern is to move the 'private' variables and everyone using them inside the **init**.

File: jquery.flot.hover.js [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+        if (options.grid.hoverable || options.grid.clickable) {
+            plot.hooks.bindEvents.push(bindEvents);
+            plot.hooks.shutdown.push(shutdown);
+            plot.hooks.drawOverlay.push(drawOverlay);
+        }
+
+        var triggerCleanupEvent = function(e) {
+            plot.unhighlight();
+            plot.getPlaceholder().trigger(new CustomEvent('plothovercleanup', { detail: e }));
```
Comment (cristiingineru) [inline jquery.flot.hover.js]: Maybe 'hoverleave'?

File: jquery.flot.hover.js [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+                highlight(item.series, item.datapoint, eventname);
+            }
+        }
+
+        plot.getPlaceholder().trigger(eventname, [pos, item]);
+    }
+
+    function highlight(s, point, auto) {
+        if (typeof s === "number") {
+            s = series[s];
```
Comment (cristiingineru) [inline jquery.flot.hover.js]: Where is the **series** variable coming from? Was this function copied from another function were the **series** variable was shared? Is this **if** statement really needed?

File: jquery.flot.hover.js [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+    function unhighlight(s, point) {
+        if (s == null && point == null) {
+            highlights = [];
+            plot.triggerRedrawOverlay();
+            return;
+        }
+
+        if (typeof s === "number") {
+            s = series[s];
```
Comment (cristiingineru) [inline jquery.flot.hover.js]: **series** again

File: jquery.flot.hover.js [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        var i = indexOfHighlight(s, point);
+        if (i !== -1) {
+            highlights.splice(i, 1);
+
+            plot.triggerRedrawOverlay();
+        }
+    }
+
+    function indexOfHighlight(s, p) {
+        for (var i = 0; i < highlights.length; ++i) {
```
Comment (cristiingineru) [inline jquery.flot.hover.js]: Is there a reason to skip the first element? Maybe a **.filter** would make this cleaner :-)

File: tests/jquery.flot.touch.Test.js [hunk 8]
```diff
@@ -401,4 +401,67 @@ describe("flot touch plugin", function () {
             expect(spy).not.toHaveBeenCalled();
         });
     });
+
+    describe('tap', function() {
```
Comment (cristiingineru) [inline tests/jquery.flot.touch.Test.js]: I wish to see a test which take into account the pan interaction as well. Basically, are the pan events expected to happen while tapping? Either way, this should be 'documented' with a test.

File: tests/jquery.flot.hover.Test.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
+describe("flot hover plugin", function () {
```
Comment (cristiingineru) [inline tests/jquery.flot.hover.Test.js]: I think this is a good start for a fist pass. Next PR I expect to see more corner cases :-P

File: jquery.flot.hover.js (hunk 7 above)
Comment (ralucaportase) [reply jquery.flot.hover.js]: This was a copy-paste from flot :)

File: jquery.flot.hover.js [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+        function shutdown(plot, eventHolder) {
+            eventHolder[0].removeEventListener('tap', tap.generatePlothoverEvent);
+            eventHolder[0].addEventListener('tap', triggerCleanupEvent);
+            eventHolder.unbind("mousemove", onMouseMove);
+            eventHolder.unbind("mouseleave", onMouseLeave);
+            eventHolder.unbind("click", onClick);
+            highlights = [];
+        }
+
+        var tap = {
```
[the remark this answers is not in the document]
Comment (ralucaportase) [reply jquery.flot.hover.js]: it is actually outside bindEvents

File: jquery.flot.hover.js (hunk 5 above)
Comment (ralucaportase) [reply jquery.flot.hover.js]: Yes, this was copy-pasted. But I think we should move it inside init and keep the if statement. The highlight and unhighlight are part of the public plot API

File: jquery.flot.hover.js (hunk 7 above)
Comment (cristiingineru) [reply jquery.flot.hover.js]: It doesn't look like :-P
